### DISCUSSION TITLE
How to propagate package dependencies to consumers?

### DISCUSSION THREAD

**Discussion by braxtons12** [xmake-io/xmake#2595] [Q&A]
I have a project that requires several packages (doctest, fmt, gsl, boost, conan::tracy).

When I package this project, how do I forward or include those dependencies so consuming projects don't have to manually add my package's dependencies to their project?

**Comment by waruqi** [comment]
`xmake install -o ./xxx`

It will copy all so/dll files if they are shared.

**Reply by braxtons12** [reply]
Is there not a way to do this without running external commands?
I feel that requiring a package as a dependency should automatically pull in all of the packages that dependency relies on as well.
If that doesn't happen, or there's not a simple way to do that from the xmake.lua, it defeats the point of having dependency resolution at all?

as an example, if I have a project with:

```
project("myproject")

add_requires("doctset", "fmt", {system = false})

target(mylib)
    set_kind("static")
    add_includedirs("include")
    add_files("src/*.cpp")
    add_packages("doctest", "fmt")
```

and another project that depends on it:
```
project("MyOtherProject")

add_repositories("local-repo path/to/myproject/build")
add_requires("myproject")

target("mybin")
    set_kind("binary")
    add_includedirs("include")
    add_files("src/*.cpp")
    add_packages("myproject")
```

Why aren't the "doctest" and "fmt" dependencies automatically propagated to "mybin"?

**Reply by braxtons12** [reply]
Bumping this since I haven't gotten a reply to my last reply yet.

If 
(1) I have a project setup like this:

```
project("MyProject")

add_requires("conan::fmt/8.1.1", { alias = "fmt"})
add_requires("conan::boost/1.79.0", {alias = "boost"})

target("MyProject"))
    set_kind("static")
    add_includedirs("include")
    add_files("src/*.cpp")
    add_packages("fmt", "boost", {public = true})
```

and
(2) I then run `xmake package` for that project

and
(3) I have another project that depends on the previous one, setup like this:

```
project("MyOtherProject")

add_repositories("local-repo path/to/MyProject/build")
add_requires("MyProjectt")

target("MyOtherProject")
    set_kind("binary")
    add_includedirs("include")
    add_files("src/*.cpp")
    add_packages("MyProject")
```
    
What do I need to add to the "MyProject" project (either the project xmake.lua or the package xmake.lua generated by `xmake package`) so that "MyOtherProject" automatically pulls in the dependencies of "MyProject" when I add "MyProject" as a dependency?

**Comment by waruqi** [comment]
> Is there not a way to do this without running external commands?
I feel that requiring a package as a dependency should automatically pull in all of the packages that dependency relies on as well.

you can only run `xmake install` to export all package deps files (`.so`, `.dll`) now when these package are shared libraries. `add_requires("fmt", {configs = {shared = true}})`

> add_requires("conan::fmt/8.1.1", { alias = "fmt"})

only builtin xmake/conda/vcpkg/cargo/cmake/apt package files will be exported, conan is not supported not.

**Reply by waruqi** [reply]
> (2) I then run xmake package for that project

It just package target and it's deps, will not package all packages.

**Reply by braxtons12** [reply]
> only builtin xmake/conda/vcpkg/cargo/cmake/apt package files will be exported, conan is not supported not.

Will exporting package files from conan packages be supported in the future?

**Reply by waruqi** [reply]
maybe, but I don't know how to get libfiles from conan. https://github.com/xmake-io/xmake/blob/master/xmake/modules/package/manager/conan/find_package.lua need be improved.

like this. https://github.com/xmake-io/xmake/blob/be3bab756494bd895943d115b05d0c32c6d42128/xmake/modules/package/manager/vcpkg/find_package.lua#L135

you can also open a pr to improve it.

btw, why not use `fmt` and `boost` in xmake-repo?

**Reply by braxtons12** [reply]
> maybe, but I don't know how to get libfiles from conan.

That's unfortunate.
 
> btw, why not use `fmt` and `boost` in xmake-repo?

I need other packages that the xmake-repo doesn't have but conan does (for example, [tracy](https://github.com/wolfpld/tracy) ) and it's simpler to keep everything from the same source in my opinion.

I would use vcpkg, since you say it is better supported, but I haven't been able to get xmake to work with it either.
(for example, trying to require any version of vcpkg::tracy in my project results in either "fetch vcpkg::tracy failed!" or "error: the baseline does not contain an entry for port tracy"

**Reply by waruqi** [reply]
You can use a mixture of them and just get the missing packages from vcpkg/conan. You can also commit the missing packages to the xmake-repo repository.

**Reply by braxtons12** [reply]
> You can use a mixture of them and just get the missing packages from vcpkg/conan.

I understand that, I would just prefer to use the same source for everything if/when possible

**Reply by waruqi** [reply]
so you can submit missing packages to xmake-repo

**Comment by waruqi** [comment]
I will try improve it. https://github.com/xmake-io/xmake/issues/2725
